logstash-core/lib/logstash/output_delegator.rb

```diff
@@ -1,192 +1,60 @@
-    @workers.subList(1,@workers.size).each(&:register)
-    setup_multi_receive!
+    @strategy.register
   end
 
-  def setup_additional_workers!(target_worker_count)
-    warn_on_worker_override!
-
-    (target_worker_count - 1).times do
-      inst = @klass.new(@config)
-      inst.metric = @metric
-      @workers << inst
-    end
-
-    # This queue is used to manage sharing across threads
-    @worker_queue = SizedQueue.new(target_worker_count)
-    @workers.each {|w| @worker_queue << w }
-  end
-
-  def setup_multi_receive!
-    # One might wonder why we don't use something like
-    # define_singleton_method(:multi_receive, method(:threadsafe_multi_receive)
-    # and the answer is this is buggy on Jruby 1.7.x . It works 98% of the time!
-    # The other 2% you get weird errors about rebinding to the same object
-    # Until we switch to Jruby 9.x keep the define_singleton_method parts
-    # the way they are, with a block
-    # See https://github.com/jruby/jruby/issues/3582
-    if threadsafe?
-      @threadsafe_worker = @workers.first
-      define_singleton_method(:multi_receive) do |events|
-        threadsafe_multi_receive(events)
-      end
-    else
-      define_singleton_method(:multi_receive) do |events|
-        worker_multi_receive(events)
-      end
-    end
-  end
-
-  def threadsafe_multi_receive(events)
-    @events_received.increment(events.length)
+  def multi_receive(events)
     @metric_events.increment(:in, events.length)
-
-    clock = @metric_events.time(:duration_in_millis)
-    @threadsafe_worker.multi_receive(events)
-    clock.stop
+    @strategy.multi_receive(events)
     @metric_events.increment(:out, events.length)
   end
 
-  def worker_multi_receive(events)
-    @events_received.increment(events.length)
-    @metric_events.increment(:in, events.length)
-
-    worker = @worker_queue.pop
-    begin
-      clock = @metric_events.time(:duration_in_millis)
-      worker.multi_receive(events)
-      clock.stop
-      @metric_events.increment(:out, events.length)
-    ensure
-      @worker_queue.push(worker)
-    end
-  end
-
   def do_close
-    @logger.debug("closing output delegator", :klass => @klass.name)
-
-    if @threadsafe
-      @workers.each(&:do_close)
-    else
-      worker_count.times do
-        worker = @worker_queue.pop
-        worker.do_close
-      end
-    end
+    @strategy.do_close
   end
-
-  def events_received
-    @events_received.value
-  end
-
-  # There's no concept of 'busy' workers for a threadsafe plugin!
-  def busy_workers
-    if @threadsafe
-      0
-    else
-      # The pipeline reporter can run before the outputs are registered trying to pull a value here
-      # In that case @worker_queue is empty, we just return 0
-      return 0 unless @worker_queue
-      @workers.size - @worker_queue.size
-    end
-  end
-
-  def worker_count
-    @workers.size
-  end
-
-  private
-  # Needed for testing, so private
-  attr_reader :threadsafe_worker, :worker_queue
-end end
+end; end
```

Comment:
Good point on only using the ID method. In my next commit I'll actually require the ID be set on plugin create. It's a sane requirement and reduces mystery.

Hmmm, @jordansissel and I arrived at the term `@strategy` after a discussion. The thing is that at some point you an `OutputDelegator` and an `OutputProxy/OutputDelegate` AND an `Output` and that all gets very confusing in terms of hierarchy.

As is you say "We have an output delegator, it has a strategy, the strategy controls the output itself", which, to my ears is a bit easier to understand.